exporter/loadbalancingexporter/resolver_dns.go

```diff
@@ -0,0 +1,171 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package loadbalancingexporter
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sort"
+	"sync"
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+	"go.uber.org/zap"
+)
+
+var _ resolver = (*dnsResolver)(nil)
+
+const (
+	defaultResInterval = 5 * time.Second
+	defaultResTimeout  = time.Second
+)
+
+var errNoHostname = errors.New("no hostname specified to resolve the backends")
+
+type dnsResolver struct {
+	logger *zap.Logger
+
+	hostname    string
+	resolver    netResolver
+	resInterval time.Duration
+	resTimeout  time.Duration
+
+	endpoints         []string
+	onChangeCallbacks []func([]string)
+
+	stopCh     chan (struct{})
+	updateLock sync.Mutex
+	shutdownWg sync.WaitGroup
+}
+
+type netResolver interface {
+	LookupIPAddr(ctx context.Context, host string) ([]net.IPAddr, error)
+}
+
+func newDNSResolver(logger *zap.Logger, hostname string) (*dnsResolver, error) {
+	if len(hostname) == 0 {
+		return nil, errNoHostname
+	}
+
+	return &dnsResolver{
+		logger:      logger,
+		hostname:    hostname,
+		resolver:    &net.Resolver{},
+		resInterval: defaultResInterval,
+		resTimeout:  defaultResTimeout,
+		stopCh:      make(chan struct{}),
+	}, nil
+}
+
+func (r *dnsResolver) start(ctx context.Context) error {
+	if _, err := r.resolve(ctx); err != nil {
+		return err
+	}
+
+	go r.periodicallyResolve()
+
+	return nil
+}
+
+func (r *dnsResolver) shutdown(ctx context.Context) error {
+	close(r.stopCh)
+	r.shutdownWg.Wait()
```

Comment:
Let's say shutdown() is called when resolve() is about to execute(). If resolve() is on line 105, but not yet executed it, then here line 87 will not wait for anything since shutdownWg counter is 0. shutdown() will return without waiting, then resolve() will be successfully executed and will call onChange callbacks, which is a violation of the resolver interface contract.
I may be wrong in the analysis, please double check.

Reply:
You are correct in the analysis. It's a fairly low probability of happening, as this is not in the critical path and two concurrent registers have to be at the exact sweet spot for this to happen. In the typical case, the side-effect also won't matter much, as the resolver will only shutdown when the exporter is shutting down, which in turn only happens when the whole collector is shutting down. This becomes a relevant problem once we have the ability to start/stop components on demand, independent on the collector's life cycle.

In any case, we have two possible solutions here:

1. clear the callbacks on shutdown (my preferred)
1. share a lock instead of wait group.

I can't find a drawback on the first solution, so, I'm adopting it as it's easier to read, understand and reason about.

Reply:
>as the resolver will only shutdown when the exporter is shutting down, which in turn only happens when the whole collector is shutting down.

It can also happen during hot reload on config change that we plan to add in the future so we need to support this cleanly.

>clear the callbacks on shutdown (my preferred)

Sounds good.